adding properties mapping parameter docs (#9572)
opensearch-project/documentation-website: _field-types/mapping-parameters/properties.md

---
layout: default
title: Properties
parent: Mapping parameters
grand_parent: Mapping and field types
nav_order: 150
has_children: false
has_toc: false
---

# Properties

The `properties` mapping parameter is used to define the structure and data types of fields within an object or the root of a document. It acts as the core of any mapping definition, allowing you to explicitly specify field names, types (such as `text`, `keyword`, `date`, or `float`), and additional settings or mapping parameters for each field.

By using `properties`, you gain full control over how your data is indexed and stored, enabling precise search behavior, aggregation support, and data validation.

## Defining fields with properties

> The following request creates an index named `products` with a structured mapping using the `properties` parameter. It includes a nested object field called `dimensions` with sub-fields:

Reviewer: ```suggestion
The following request creates an index named `products` with a structured mapping using the `properties` parameter. It includes a nested object field called `dimensions` with subfields:
```

```json
PUT /products
{
  "mappings": {
    "properties": {
      "name": {
        "type": "text"
      },
      "sku": {
        "type": "keyword"
      },
      "price": {
        "type": "float"
      },
      "available": {
        "type": "boolean"
      },
      "created_at": {
        "type": "date",
        "format": "yyyy-MM-dd"
      },
      "dimensions": {
        "type": "object",
        "properties": {
          "width": { "type": "float" },
          "height": { "type": "float" },
          "depth": { "type": "float" }
        }
      }
    }
  }
}
```
{% include copy-curl.html %}

## Indexing a document

> Use the following command to index a document with nested fields:

Reviewer 2: Could you link to the nested fields page?

```json
PUT /products/_doc/1
{
  "name": "Wireless Mouse",
  "sku": "WM-1001",
  "price": 24.99,
  "available": true,
  "created_at": "2024-12-01",
  "dimensions": {
    "width": 6.5,
    "height": 3.2,
    "depth": 1.5
  }
}
```
{% include copy-curl.html %}

## Querying and aggregating using dot notation

> You can query or aggregate on sub-fields inside objects using dot notation. Use the following command to execute a query that:

Reviewer 2: ```suggestion
You can query or aggregate on object subfields using dot notation. Use the following command to execute a query that:
```

> - Uses `dimensions.width` field to filter documents where `width` is between `5` and `10`.

Reviewer 2: ```suggestion
- Filters documents on the `dimensions.width` field, returning documents in which `width` is between `5` and `10`.
```

> - Creates [histogram aggregation]({{site.url}}{{site.baseurl}}/aggregations/bucket/histogram/) on `dimensions.depth` field which creates buckets for products `depth` using intervals of `0.5`.

Reviewer 2: ```suggestion
- Creates a [histogram aggregation]({{site.url}}{{site.baseurl}}/aggregations/bucket/histogram/) on the `dimensions.depth` field, creating buckets for products using `depth` intervals of `0.5`.
```

```json
POST /products/_search
{
  "query": {
    "range": {
      "dimensions.width": {
        "gte": 5,
        "lte": 10
      }
    }
  },
  "aggs": {
    "Depth Distribution": {
      "histogram": {
        "field": "dimensions.depth",
        "interval": 0.5
      }
    }
  }
}
```
{% include copy-curl.html %}

> Expected result:

Reviewer 2: Could you add a more descriptive sentence for the response?

> The following response shows a matching document where the dimensions.width field falls within the specified range. It also includes a histogram aggregation result for dimensions.depth:

Reviewer: ```suggestion
The following response shows a matching document in which the `dimensions.width` field falls within the specified range. It also includes a histogram aggregation result for `dimensions.depth`:
```
